File: part1/hs/ex1.hs
maybeCompare :: (Eq a) => Maybe Int -> a -> (a, Int) -> Maybe Int
maybeCompare m x (y, i)
    | x == y = Just i
    | otherwise = m

tuplelist :: [a] -> [(a, Int)]
tuplelist xs = map (\x -> (xs !! x, x)) [0,1..(length xs) - 1]

extract :: [a] -> Maybe Int -> Maybe a
extract xs Nothing = Nothing
extract xs (Just i)
    | (i + 1) < (length xs) = Just $ xs !! (i + 1)
    | otherwise = Nothing

next :: (Eq a) => [a] -> a -> Maybe a
next xs x = extract xs $ foldl ((flip maybeCompare) x) Nothing $ tuplelist xs

File: part1/js/ex1.js
// [a] -> [(a, Int)]
function toTuple(list) {
    return list.map(function (x, idx) {
        return [x, idx];
    });
}

/*
 * Return the item next to the specified item in a list.
 * If the item doesn't exist, return undefined.
 * If there is more than one item, use the last found item.
 */

// (Eq a) => [a] -> a -> Maybe a
function next(list, item) {
    tupleList = toTuple(list);

    idx = tupleList.reduce(function (old, current) {
        if(current[0] === item) {
            old = current[1] + 1;
        }

        return old;
    }, undefined);

    if(idx === list.length || idx === undefined) {
        return undefined;
    } else {
        return list[idx];
    }
}

module.exports = next;
